### PR TITLE
Optimize quantization kernels

### PR DESCRIPTION
**Pull request by vshampor**
Switched to using 2D grid for per-channel kernels and cleaned up
the code duplications across the extensions.

### CODE REVIEW

**Comment by vshampor** [comment]
Internal benchmark results:

```
CUDA low batch
------------------------------------------------
Custom Symmetric (cuda 0 ) impl:
input size: [2, 96, 112, 112]
Before:
Forward: min 0.195ms / avg 0.350ms | Backward: min 0.232ms / avg 0.561ms
After:
Forward: min 0.196ms / avg 0.354ms | Backward: min 0.235ms / avg 0.535ms

Custom Symmetric Per Weight Channel  (cuda 0 ) impl
input size: [2, 96, 112, 112]
Before:
Forward: min 0.197ms / avg 0.379ms | Backward: min 0.249ms / avg 0.650ms
After:
Forward: min 0.196ms / avg 0.356ms | Backward: min 0.249ms / avg 0.556ms

Custom Symmetric Per Activation Channel  (cuda 0 ) impl
input size: [2, 96, 112, 112]
Before:
Forward: min 0.198ms / avg 0.417ms | Backward: min 1.004ms / avg 1.690ms
After:
Forward: min 0.197ms / avg 0.407ms | Backward: min 0.266ms / avg 0.692ms

Custom Asymmetric (cuda 0 ) impl:
input size: [2, 96, 112, 112]
Before:
Forward: min 0.618ms / avg 0.862ms | Backward: min 0.254ms / avg 0.595ms
After:
Forward: min 0.624ms / avg 0.864ms | Backward: min 0.239ms / avg 0.564ms

Custom Asymmetric Per Weight Channel  (cuda 0 ) impl
input size: [2, 96, 112, 112]
Before:
Forward: min 0.633ms / avg 0.905ms | Backward: min 0.268ms / avg 0.640ms
After:
Forward: min 0.634ms / avg 0.901ms | Backward: min 0.249ms / avg 0.618ms

Custom Asymmetric Per Activation Channel  (cuda 0 ) impl
input size: [2, 96, 112, 112]
Before:
Forward: min 0.623ms / avg 1.571ms | Backward: min 1.008ms / avg 1.919ms
After:
Forward: min 0.633ms / avg 0.897ms | Backward: min 0.278ms / avg 0.576ms

CUDA high batch
------------------------------------------------
Custom Symmetric (cuda 0 ) impl:
input size: [128, 96, 112, 112]
Before:
Forward: min 3.755ms / avg 3.804ms | Backward: min 6.175ms / avg 6.269ms
After:
Forward: min 3.782ms / avg 3.864ms | Backward: min 6.187ms / avg 6.288ms


Custom Symmetric Per Weight Channel  (cuda 0 ) impl
input size: [128, 96, 112, 112]
Before:
Forward: min 3.787ms / avg 3.801ms | Backward: min 5.844ms / avg 5.913ms
After:
Forward: min 3.779ms / avg 3.848ms | Backward: min 5.955ms / avg 6.052ms



Custom Symmetric Per Activation Channel  (cuda 0 ) impl
input size: [128, 96, 112, 112]
Before:
Forward: min 3.801ms / avg 3.846ms | Backward: min 6.083ms / avg 6.194ms
After:
Forward: min 3.783ms / avg 3.836ms | Backward: min 6.521ms / avg 6.604ms

Custom Asymmetric (cuda 0 ) impl:
input size: [128, 96, 112, 112]
Before:
Forward: min 4.198ms / avg 4.297ms | Backward: min 6.135ms / avg 6.288ms
After:
Forward: min 4.238ms / avg 4.385ms | Backward: min 6.127ms / avg 6.274ms


Custom Asymmetric Per Weight Channel  (cuda 0 ) impl
input size: [128, 96, 112, 112]
Before:
Forward: min 4.206ms / avg 4.317ms | Backward: min 5.938ms / avg 6.086ms
After:
Forward: min 4.241ms / avg 4.392ms | Backward: min 5.890ms / avg 6.016ms

Custom Asymmetric Per Activation Channel  (cuda 0 ) impl
input size: [128, 96, 112, 112]
Before:
Forward: min 4.227ms / avg 4.282ms | Backward: min 6.107ms / avg 6.183ms
After:
Forward: min 4.244ms / avg 4.356ms | Backward: min 6.545ms / avg 6.636ms
```

While at high batches the improvements are not quite visible, for lower batches the speedup for the per-activation kernel version is drastic.

**Comment by vshampor** [comment]
Training MobileNetV2 on CIFAR100:

With per-tensor quantization:
```
: Epoch: [0][0/196] Lr: 1e-05 Time: 0.620 (0.620) Data: 0.159 (0.159) CE_loss: 1.2441 (1.2441) CR_loss: 0.0000 (0.0000) Loss: 1.2441 (1.2441) Acc@1: 62.891 (62.891) Acc@5: 88.672 (88.672)
: Epoch: [0][10/196] Lr: 1e-05 Time: 0.152 (0.202) Data: 0.001 (0.015) CE_loss: 1.2134 (1.1613) CR_loss: 0.0000 (0.0000) Loss: 1.2134 (1.1613) Acc@1: 66.406 (67.188) Acc@5: 89.453 (89.773)
: Epoch: [0][20/196] Lr: 1e-05 Time: 0.146 (0.180) Data: 0.001 (0.008) CE_loss: 1.1982 (1.1645) CR_loss: 0.0000 (0.0000) Loss: 1.1982 (1.1645) Acc@1: 66.016 (67.857) Acc@5: 89.453 (89.286)
: Epoch: [0][30/196] Lr: 1e-05 Time: 0.168 (0.174) Data: 0.001 (0.006) CE_loss: 1.3111 (1.1602) CR_loss: 0.0000 (0.0000) Loss: 1.3111 (1.1602) Acc@1: 63.281 (68.107) Acc@5: 85.938 (89.176)
: Epoch: [0][40/196] Lr: 1e-05 Time: 0.160 (0.170) Data: 0.001 (0.005) CE_loss: 1.1461 (1.1682) CR_loss: 0.0000 (0.0000) Loss: 1.1461 (1.1682) Acc@1: 68.750 (67.978) Acc@5: 87.109 (88.929)
: Epoch: [0][50/196] Lr: 1e-05 Time: 0.144 (0.168) Data: 0.001 (0.004) CE_loss: 1.1719 (1.1680) CR_loss: 0.0000 (0.0000) Loss: 1.1719 (1.1680) Acc@1: 66.016 (68.076) Acc@5: 89.062 (88.932)
: Epoch: [0][60/196] Lr: 1e-05 Time: 0.156 (0.166) Data: 0.001 (0.003) CE_loss: 1.1661 (1.1727) CR_loss: 0.0000 (0.0000) Loss: 1.1661 (1.1727) Acc@1: 68.359 (67.943) Acc@5: 88.672 (88.813)
: Epoch: [0][70/196] Lr: 1e-05 Time: 0.127 (0.162) Data: 0.001 (0.003) CE_loss: 1.1123 (1.1634) CR_loss: 0.0000 (0.0000) Loss: 1.1123 (1.1634) Acc@1: 67.969 (67.991) Acc@5: 89.062 (88.958)
: Epoch: [0][80/196] Lr: 1e-05 Time: 0.128 (0.160) Data: 0.001 (0.003) CE_loss: 1.1186 (1.1628) CR_loss: 0.0000 (0.0000) Loss: 1.1186 (1.1628) Acc@1: 73.438 (68.113) Acc@5: 89.453 (88.942)
: Epoch: [0][90/196] Lr: 1e-05 Time: 0.150 (0.159) Data: 0.001 (0.003) CE_loss: 1.2279 (1.1656) CR_loss: 0.0000 (0.0000) Loss: 1.2279 (1.1656) Acc@1: 66.406 (68.012) Acc@5: 87.891 (88.908)
: Epoch: [0][100/196] Lr: 1e-05 Time: 0.144 (0.160) Data: 0.001 (0.002) CE_loss: 1.0551 (1.1615) CR_loss: 0.0000 (0.0000) Loss: 1.0551 (1.1615) Acc@1: 70.312 (68.197) Acc@5: 92.578 (89.032)
: Epoch: [0][110/196] Lr: 1e-05 Time: 0.152 (0.159) Data: 0.001 (0.002) CE_loss: 1.2757 (1.1600) CR_loss: 0.0000 (0.0000) Loss: 1.2757 (1.1600) Acc@1: 64.062 (68.226) Acc@5: 86.719 (89.052)
```

With per-channel quantization (before optimizations):
```
:Epoch: [0][0/196] Lr: 1e-05 Time: 1.437 (1.437) Data: 0.159 (0.159) CE_loss: 5.4935 (5.4935) CR_loss: 0.0000 (0.0000) Loss: 5.4935 (5.4935) Acc@1: 7.812 (7.812) Acc@5: 24.609 (24.609)
: Epoch: [0][10/196] Lr: 1e-05 Time: 1.022 (1.041) Data: 0.001 (0.015) CE_loss: 5.4856 (5.2987) CR_loss: 0.0000 (0.0000) Loss: 5.4856 (5.2987) Acc@1: 8.984 (9.197) Acc@5: 22.656 (26.172)
: Epoch: [0][20/196] Lr: 1e-05 Time: 0.972 (1.020) Data: 0.001 (0.008) CE_loss: 5.4675 (5.2672) CR_loss: 0.0000 (0.0000) Loss: 5.4675 (5.2672) Acc@1: 8.203 (9.319) Acc@5: 29.688 (27.083)
: Epoch: [0][30/196] Lr: 1e-05 Time: 1.015 (1.008) Data: 0.001 (0.006) CE_loss: 5.1102 (5.1890) CR_loss: 0.0000 (0.0000) Loss: 5.1102 (5.1890) Acc@1: 7.422 (9.866) Acc@5: 27.734 (28.201)
: Epoch: [0][40/196] Lr: 1e-05 Time: 1.003 (0.998) Data: 0.001 (0.005) CE_loss: 4.7161 (5.1260) CR_loss: 0.0000 (0.0000) Loss: 4.7161 (5.1260) Acc@1: 12.500 (10.232) Acc@5: 35.156 (28.887)
: Epoch: [0][50/196] Lr: 1e-05 Time: 0.960 (0.997) Data: 0.001 (0.004) CE_loss: 4.9161 (5.0583) CR_loss: 0.0000 (0.0000) Loss: 4.9161 (5.0583) Acc@1: 12.500 (10.409) Acc@5: 29.688 (29.450)
```

With per-channel quantization (after optimizations):
```
: Epoch: [0][0/196] Lr: 1e-05 Time: 0.583 (0.583) Data: 0.182 (0.182) CE_loss: 5.3470 (5.3470) CR_loss: 0.0000 (0.0000) Loss: 5.3470 (5.3470) Acc@1: 10.156 (10.156) Acc@5: 25.391 (25.391)
: Epoch: [0][10/196] Lr: 1e-05 Time: 0.163 (0.201) Data: 0.001 (0.017) CE_loss: 4.9830 (5.1725) CR_loss: 0.0000 (0.0000) Loss: 4.9830 (5.1725) Acc@1: 11.719 (9.943) Acc@5: 28.906 (27.024)
: Epoch: [0][20/196] Lr: 1e-05 Time: 0.166 (0.181) Data: 0.001 (0.009) CE_loss: 5.1160 (5.1468) CR_loss: 0.0000 (0.0000) Loss: 5.1160 (5.1468) Acc@1: 8.203 (9.542) Acc@5: 28.516 (27.269)
: Epoch: [0][30/196] Lr: 1e-05 Time: 0.150 (0.175) Data: 0.001 (0.007) CE_loss: 5.3542 (5.1315) CR_loss: 0.0000 (0.0000) Loss: 5.3542 (5.1315) Acc@1: 8.594 (9.627) Acc@5: 26.562 (28.049)
: Epoch: [0][40/196] Lr: 1e-05 Time: 0.128 (0.164) Data: 0.001 (0.005) CE_loss: 4.5618 (5.0851) CR_loss: 0.0000 (0.0000) Loss: 4.5618 (5.0851) Acc@1: 16.797 (10.051) Acc@5: 35.938 (28.420)
: Epoch: [0][50/196] Lr: 1e-05 Time: 0.131 (0.159) Data: 0.001 (0.004) CE_loss: 4.6659 (5.0224) CR_loss: 0.0000 (0.0000) Loss: 4.6659 (5.0224) Acc@1: 12.500 (10.432) Acc@5: 34.766 (29.282)
: Epoch: [0][60/196] Lr: 1e-05 Time: 0.129 (0.154) Data: 0.001 (0.004) CE_loss: 4.8114 (4.9708) CR_loss: 0.0000 (0.0000) Loss: 4.8114 (4.9708) Acc@1: 13.281 (10.713) Acc@5: 34.766 (29.867)
: Epoch: [0][70/196] Lr: 1e-05 Time: 0.127 (0.150) Data: 0.001 (0.003) CE_loss: 4.8823 (4.9388) CR_loss: 0.0000 (0.0000) Loss: 4.8823 (4.9388) Acc@1: 12.500 (10.849) Acc@5: 30.859 (30.194)
: Epoch: [0][80/196] Lr: 1e-05 Time: 0.128 (0.148) Data: 0.001 (0.003) CE_loss: 4.6017 (4.8926) CR_loss: 0.0000 (0.0000) Loss: 4.6017 (4.8926) Acc@1: 12.891 (11.111) Acc@5: 32.031 (30.710)
: Epoch: [0][90/196] Lr: 1e-05 Time: 0.130 (0.146) Data: 0.001 (0.003) CE_loss: 4.6174 (4.8549) CR_loss: 0.0000 (0.0000) Loss: 4.6174 (4.8549) Acc@1: 11.719 (11.247) Acc@5: 30.859 (31.095)
: Epoch: [0][100/196] Lr: 1e-05 Time: 0.127 (0.145) Data: 0.001 (0.003) CE_loss: 4.2377 (4.8210) CR_loss: 0.0000 (0.0000) Loss: 4.2377 (4.8210) Acc@1: 14.062 (11.394) Acc@5: 35.156 (31.451)
: Epoch: [0][110/196] Lr: 1e-05 Time: 0.133 (0.143) Data: 0.001 (0.002) CE_loss: 4.2912 (4.7884) CR_loss: 0.0000 (0.0000) Loss: 4.2912 (4.7884) Acc@1: 12.891 (11.501) Acc@5: 37.891 (31.715)
: Epoch: [0][120/196] Lr: 1e-05 Time: 0.128 (0.142) Data: 0.001 (0.002) CE_loss: 4.1432 (4.7521) CR_loss: 0.0000 (0.0000) Loss: 4.1432 (4.7521) Acc@1: 13.281 (11.709) Acc@5: 38.281 (31.970)
: Epoch: [0][130/196] Lr: 1e-05 Time: 0.151 (0.141) Data: 0.001 (0.002) CE_loss: 4.1050 (4.7146) CR_loss: 0.0000 (0.0000) Loss: 4.1050 (4.7146) Acc@1: 15.234 (12.011) Acc@5: 38.672 (32.374)
: Epoch: [0][140/196] Lr: 1e-05 Time: 0.125 (0.140) Data: 0.001 (0.002) CE_loss: 4.1902 (4.6746) CR_loss: 0.0000 (0.0000) Loss: 4.1902 (4.6746) Acc@1: 12.891 (12.201) Acc@5: 36.328 (32.804)
: Epoch: [0][150/196] Lr: 1e-05 Time: 0.150 (0.140) Data: 0.001 (0.002) CE_loss: 4.1070 (4.6395) CR_loss: 0.0000 (0.0000) Loss: 4.1070 (4.6395) Acc@1: 16.797 (12.376) Acc@5: 36.719 (33.177)
: Epoch: [0][160/196] Lr: 1e-05 Time: 0.128 (0.140) Data: 0.001 (0.002) CE_loss: 3.8454 (4.6060) CR_loss: 0.0000 (0.0000) Loss: 3.8454 (4.6060) Acc@1: 17.578 (12.566) Acc@5: 44.922 (33.565)
: Epoch: [0][170/196] Lr: 1e-05 Time: 0.167 (0.141) Data: 0.001 (0.002) CE_loss: 3.9928 (4.5710) CR_loss: 0.0000 (0.0000) Loss: 3.9928 (4.5710) Acc@1: 17.578 (12.701) Acc@5: 38.281 (33.950)
: Epoch: [0][180/196] Lr: 1e-05 Time: 0.163 (0.142) Data: 0.001 (0.002) CE_loss: 3.8937 (4.5424) CR_loss: 0.0000 (0.0000) Loss: 3.8937 (4.5424) Acc@1: 16.016 (12.850) Acc@5: 41.406 (34.278)
: Epoch: [0][190/196] Lr: 1e-05 Time: 0.162 (0.143) Data: 0.000 (0.002) CE_loss: 4.1302 (4.5137) CR_loss: 0.0000 (0.0000) Loss: 4.1302 (4.5137) Acc@1: 14.844 (13.034) Acc@5: 34.766 (34.582)
```

With optimized kernels, the per-channel quantization training has become 7x faster and is now on par with the regular per-tensor quantization.

File: nncf/extensions/include/common_cuda_defs.cuh [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    }   \
+    cudaDeviceSynchronize(); \
+    auto start = std::chrono::steady_clock::now(); \
+    for (int i = 0; i < iter; i++) {   \
+            CODE                          \
+        }   \
+    cudaDeviceSynchronize(); \
+    auto end = std::chrono::steady_clock::now(); \
+    std::chrono::duration<double> diff = (end - start) / iter; \
+    std::cout << "VSHAMPOR: time = " << std::chrono::duration_cast<std::chrono::nanoseconds>(diff).count() << std::endl; \
```
Comment (ljaljushkin) [inline nncf/extensions/include/common_cuda_defs.cuh:60]: I think name shouldn't appear even in c++ code

Comment (vshampor) [reply nncf/extensions/include/common_cuda_defs.cuh:60]: Sorry, should have deleted it.

**Comment by vshampor** [comment]
Actually no, `reduce_with_shared_memory` is just a new, more fitting name for the `sumGrad` function. The optimizations are with the new kernel launch configuration for the per-channel kernels - the `grid_size` kernel launch argument in the per-channel kernels is now 2D (X is for channel, Y is for the rest of the blocks that "fit" into the GPU for a given block size in terms of threads). Previously multiple scales were handled in a `for` loop of 1d (X for all available blocks on GPU for a given block thread size) kernel launches, where each iteration was responsible for its own scale channel.

**Comment by vanyalzr** [comment]
@vshampor Results for the benchmark runs of the kernel for a batch of 128 samples show no speedup, whereas training (with a large batch) shows a speedup. Are the benchmark results outdated or what you think is a result of this discrepancy?

**Comment by vshampor** [comment]
@vanyalzr that's a mystery for me as well. Going through the PyTorch extension API to do a layer benchmark might involve all kinds of unaccounted-for effects - that is why, during my optimization experiments, I used a more local benchmarking approach that directly measures kernel execution times (used a custom C++ app and the `PROFILE` macro I introduced with this commit). These showed clear improvements even in the large-batch scenario (didn't save the logs, though, but may do a re-run)

**Pull request merged**